Infrastructure/API Support for Composite Uploads/Downloads

jmchilton opened the pull request:
Extends the upload tool to allow attaching arbitrary composite files to datasets on upload. Adds an API endpoint to list a dataset's composite files and directories. An API for downloading such content already exists - but the API consumer would need to know the files exist ahead of time (as is the case say with tool tests that leverage this API). Add a Directory datatype that can represent the decompressed contents of a tar dataset and the needed converter. I'll follow up with similar converters for zip files if this is merged.

jmchilton committed:
Implement API endpoint for indexing extra files in a dataset.

jmchilton committed:
Directory datatype for Galaxy.

mvdbeek commented on test/api/test_tools.py:
```diff
@@ -221,6 +221,59 @@ def test_test_data_yaml_tools(self):
+    def test_upload_composite_as_tar(self):
+        with self.dataset_populator.test_history() as history_id:
+            tar_path = self.test_data_resolver.get_filename("testdir.tar")
+            tar_f = open(tar_path, "rb")
```
We can use a `with` statement here

mvdbeek commented on test/api/test_tools.py:
```diff
@@ -221,6 +221,59 @@ def test_test_data_yaml_tools(self):
         assert len(test_data) == 3
 
+    def test_upload_composite_as_tar(self):
+        with self.dataset_populator.test_history() as history_id:
```
We could also use the history_id decorator here, right ?

mvdbeek commented on test/api/test_tools.py:
```diff
@@ -221,6 +221,59 @@ def test_test_data_yaml_tools(self):
+    def test_upload_composite_from_bad_tar(self):
+        with self.dataset_populator.test_history() as history_id:
+            tar_path = self.test_data_resolver.get_filename("unsafe.tar")
+            tar_f = open(tar_path, "rb")
```
Can use `with` here as well.

mvdbeek commented on tools/data_source/upload.py:
```diff
@@ -172,32 +174,65 @@ def add_file(dataset, registry, output_path):
 def add_composite_file(dataset, output_path, files_path):
+
+    def to_path(path_or_url):
+        is_url = path_or_url.find('://') != -1  # todo fixme
```
I suppose you still want to do something about this ?

jmchilton replied:
Well you can see the code in red below. I'm not sure this was a @jmchilton fixme - I'd have used ``TODO`` instead of ``todo`` and misspelled something.

mvdbeek replied:
how could I have missed that 😜

mvdbeek reviewed (commented):
I like it, and I think this has been requested many times before.

mvdbeek commented on lib/galaxy/datatypes/data.py:
```diff
@@ -927,6 +927,10 @@ def regex_line_dataprovider(self, dataset, **settings):
         return dataproviders.line.RegexLineDataProvider(dataset_source, **settings)
 
 
+class Directory(Text):
```
What's the reason to inherit from Text ?

jmchilton replied:
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/text.py#L24 HTML felt like the closest thing and it descends from ``Text`` also. Do you have an alternative suggestion?

mvdbeek replied:
plain Data, I think composite datatypes wouldn't have to be restricted to text files ? Or maybe they are ?

jmchilton committed:
Implement composite uploads on arbitrary datatypes.

Prior to this, composite uploads were only allowed for select data types. As far as I can tell, the framework itself allows arbitrary datatypes to have extra files. This allows the upload API to take in extra files for datatypes that aren't explicitly annotated has having extra files.

These may be specified one at a time or in directory structures via tar files.

jmchilton commented:
Tried rebasing on top of ISA-TAB changes but it seems something went awry and the ISA-TAB tests broke - moving back into WIP.

mvdbeek reviewed (approved):
This looks good to me, but I have the vague feeling `force_composite` should be documented a little more. I'm not sure what the right place for that would be, maybe as help attribute in the upload tool ?

jmchilton commented:
> I have the vague feeling force_composite should be documented a little more.

I don't think anyone can see the upload.xml help stuff, and I'm not sure I think this is the greatest entry point for general purpose usage. It allows me to do some stuff downstream in the CWL branch and might allow others to do cool stuff by attaching extra data to datasets also - but I don't think it is quite ready to advertise. If I was going to document upload API stuff I'd write up something on upload 2.0 - I think that would be a better use of time. For now I think the tests are going to be the docs, if someone asks about this at least we have something to point at. Thanks for the approve - I'm going to abuse it and just merge.
